File: solutions/c/armstrong-numbers/1/armstrong_numbers.c
#include "armstrong_numbers.h"
#include <math.h>

bool is_armstrong_number(int candidate) {
    // Handle negative numbers (Armstrong numbers are defined for non-negative integers)
    if (candidate < 0) {
        return false;
    }
    
    // Special case: 0 is an Armstrong number (0^1 = 0)
    if (candidate == 0) {
        return true;
    }
    
    // Count the number of digits
    int num = candidate;
    int digit_count = 0;
    
    while (num > 0) {
        digit_count++;
        num /= 10;
    }
    
    // Calculate sum of digits raised to power of digit_count
    num = candidate;
    long long sum = 0;  // Use long long to avoid overflow for large numbers
    
    while (num > 0) {
        int digit = num % 10;
        
        // Calculate digit^digit_count using pow() from math.h
        // Could also use a loop for integer exponentiation
        sum += (long long)pow(digit, digit_count);
        
        // Early exit if sum exceeds candidate (optimization)
        if (sum > candidate) {
            return false;
        }
        
        num /= 10;
    }
    
    return sum == candidate;
}

File: solutions/fortran/armstrong-numbers/1/armstrong_numbers.f90
module armstrong_numbers
  implicit none
contains

  logical function isArmstrongNumber(i)
    integer, intent(in) :: i
    integer :: num, digit, num_digits, sum, temp
    
    ! Handle negative numbers (not Armstrong by definition)
    if (i < 0) then
      isArmstrongNumber = .false.
      return
    end if
    
    ! Count number of digits
    num = i
    num_digits = 0
    temp = i
    
    do while (temp > 0)
      num_digits = num_digits + 1
      temp = temp / 10
    end do
    
    ! Handle single digit numbers and 0
    if (num_digits == 0) then  ! i = 0
      isArmstrongNumber = .true.  ! 0^1 = 0
      return
    end if
    
    if (num_digits == 1) then
      isArmstrongNumber = .true.  ! All single digit numbers are Armstrong numbers
      return
    end if
    
    ! Calculate sum of digits raised to power of num_digits
    sum = 0
    num = i
    
    do while (num > 0)
      digit = mod(num, 10)  ! Get last digit
      
      ! Raise digit to power of num_digits
      sum = sum + digit ** num_digits
      
      ! Remove last digit
      num = num / 10
    end do
    
    ! Check if sum equals original number
    isArmstrongNumber = (sum == i)
    
  end function isArmstrongNumber

end module armstrong_numbers